src/collider/BVHTree.cpp

```diff
@@ -158,3 +160,252 @@ BVHNode *BVHTree::AllocNode()
 		Log::Error("Out of space in m_bvhNodes.");
 	return &m_bvhNodes[m_nodeAllocPos++];
 }
+
+double BVHTree::CalculateSAH() const
+{
+	double outSAH = 0.0;
+	std::vector<BVHNode *> m_nodeStack = { m_root };
+
+	while (!m_nodeStack.empty()) {
+		BVHNode *node = m_nodeStack.back();
+		m_nodeStack.pop_back();
+
+		if (!node->IsLeaf() && node->kids[0])
+			m_nodeStack.push_back(node->kids[0]);
+		if (!node->IsLeaf() && node->kids[1])
+			m_nodeStack.push_back(node->kids[1]);
+
+		// surface area = 2 * width * depth * height
+		vector3d size = node->aabb.max - node->aabb.min;
+		double area = 2.f * size.x * size.y * size.z;
+
+		// Cost function according to https://users.aalto.fi/~laines9/publications/aila2013hpg_paper.pdf Eq. 1
+		if (node->IsLeaf())
+			outSAH += node->numTris * area;
+		else
+			outSAH += 1.2 * area;
+	}
+
+	vector3d rootSize = m_root->aabb.max - m_root->aabb.min;
+	float rootArea = 2.f * rootSize.x * rootSize.y * rootSize.z;
+
+	// Perform 1 / Aroot * ( SAH sums )
+	outSAH /= rootArea;
```

Comment:
So we just converted that all to a float (prev comment) and then are using it as the denominator for double maths?